File: app/src/main/java/com/cray/software/passwords/dialogs/GeneratePassword.java
package com.cray.software.passwords.dialogs;

import android.app.Activity;
import android.content.Intent;
import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.view.View;
import android.widget.Button;
import android.widget.SeekBar;
import android.widget.Toast;

import com.cray.software.passwords.R;
import com.cray.software.passwords.databinding.DialogGeneratePasswordBinding;
import com.cray.software.passwords.utils.ThemeUtil;

import java.util.HashMap;
import java.util.Random;

public class GeneratePassword extends Activity implements View.OnClickListener {

    private static final String AZ = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
    private static final String az = "abcdefghijklmnopqrstuvwxyz";
    private static final String numeric = "0123456789";
    private static final String symbols = "!@#$%()?^*";

    private DialogGeneratePasswordBinding binding;
    @NonNull
    private Random rnd = new Random();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        ThemeUtil themeUtil = ThemeUtil.getInstance(this);
        if (themeUtil != null) setTheme(themeUtil.getDialogStyle());
        binding = DataBindingUtil.setContentView(this, R.layout.dialog_generate_password);
        if (themeUtil != null) binding.bgView.setBackgroundColor(themeUtil.getBackgroundStyle());

        binding.passwordLength.setMax(32);
        binding.passwordLength.setProgress(16);
        binding.passLength.setText(String.valueOf(binding.passwordLength.getProgress()));
        binding.passwordLength.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                binding.passLength.setText(String.valueOf(progress));
            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {

            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {

            }
        });

        Button genButton = findViewById(R.id.genButton);
        genButton.setOnClickListener(this);
        Button genCancel = findViewById(R.id.genCancel);
        genCancel.setOnClickListener(this);
        Button genInsert = findViewById(R.id.genInsert);
        genInsert.setOnClickListener(this);
    }

    private String randomString(int len, int upper, int lower, int num, int sym) {
        String up = null;
        if (upper != 0) {
            up = AZ;
        }
        String low = null;
        if (lower != 0) {
            low = az;
        }
        String numerical = null;
        if (num != 0) {
            numerical = numeric;
        }
        String symbolical = null;
        if (sym != 0) {
            symbolical = symbols;
        }
        int mapLength = upper + lower + num + sym;
        HashMap<Integer, String> map = new HashMap<>();

        for (int i = 0; i < mapLength; i++) {
            if (up != null && !map.containsValue(up)) {
                map.put(i, up);
            }
            if (low != null && !map.containsValue(low)) {
                map.put(i, low);
            }
            if (numerical != null && !map.containsValue(numerical)) {
                map.put(i, numerical);
            }
            if (symbolical != null && !map.containsValue(symbolical)) {
                map.put(i, symbolical);
            }
        }
        StringBuilder sb = new StringBuilder(len);
        for (int i = 0; i < len; i++) {
            Random type = new Random();
            int next = type.nextInt(mapLength);
            String rand = map.get(next);
            sb.append(rand.charAt(rnd.nextInt(rand.length())));
        }
        return sb.toString();
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.genCancel:
                finish();
                break;
            case R.id.genButton:
                if (preGenerateCheck()) {
                    startGenerating();
                }
                break;
            case R.id.genInsert:
                Intent intent = new Intent();
                String genPass = binding.passwordShow.getText().toString().trim();
                intent.putExtra("GENERATED_PASSWORD", genPass);
                setResult(RESULT_OK, intent);
                finish();
                break;
        }
    }

    private boolean preGenerateCheck() {
        boolean status;
        if (binding.aZCheck.isChecked() || binding.azCheck.isChecked() || binding.numericCheck.isChecked() || binding.symbolCheck.isChecked()) {
            if (binding.passwordLength.getProgress() > 0) {
                status = true;
            } else {
                status = false;
                Toast.makeText(GeneratePassword.this, getString(R.string.pass_length), Toast.LENGTH_SHORT).show();
            }
        } else {
            status = false;
            binding.aZCheck.setChecked(true);
            Toast.makeText(GeneratePassword.this, getString(R.string.pass_checks), Toast.LENGTH_SHORT).show();
        }
        return status;
    }

    private void startGenerating() {
        int aZCh = 0;
        if (binding.aZCheck.isChecked()) {
            aZCh = 1;
        }
        int azCh = 0;
        if (binding.azCheck.isChecked()) {
            azCh = 1;
        }
        int numCh = 0;
        if (binding.numericCheck.isChecked()) {
            numCh = 1;
        }
        int symCh = 0;
        if (binding.symbolCheck.isChecked()) {
            symCh = 1;
        }
        String password = randomString(binding.passwordLength.getProgress(), aZCh, azCh, numCh, symCh);
        binding.passwordShow.setText(password);
    }
}

File: app/src/main/java/com/cray/software/passwords/dialogs/GeneratePassword.kt
package com.cray.software.passwords.dialogs

import android.content.Intent
import androidx.databinding.DataBindingUtil
import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import android.view.View
import android.widget.Button
import android.widget.SeekBar
import android.widget.Toast

import com.cray.software.passwords.R
import com.cray.software.passwords.databinding.DialogGeneratePasswordBinding
import com.cray.software.passwords.utils.ThemeUtil

import java.util.HashMap
import java.util.Random

class GeneratePassword : AppCompatActivity(), View.OnClickListener {

    private var binding: DialogGeneratePasswordBinding? = null
    private val rnd = Random()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val themeUtil = ThemeUtil.getInstance(this)
        if (themeUtil != null) setTheme(themeUtil.dialogStyle)
        binding = DataBindingUtil.setContentView<T>(this, R.layout.dialog_generate_password)
        if (themeUtil != null) binding!!.bgView.setBackgroundColor(themeUtil.backgroundStyle)

        binding!!.passwordLength.max = 32
        binding!!.passwordLength.progress = 16
        binding!!.passLength.text = binding!!.passwordLength.progress.toString()
        binding!!.passwordLength.setOnSeekBarChangeListener(object : SeekBar.OnSeekBarChangeListener {
            override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
                binding!!.passLength.text = progress.toString()
            }

            override fun onStartTrackingTouch(seekBar: SeekBar) {

            }

            override fun onStopTrackingTouch(seekBar: SeekBar) {

            }
        })

        val genButton = findViewById<Button>(R.id.genButton)
        genButton.setOnClickListener(this)
        val genCancel = findViewById<Button>(R.id.genCancel)
        genCancel.setOnClickListener(this)
        val genInsert = findViewById<Button>(R.id.genInsert)
        genInsert.setOnClickListener(this)
    }

    private fun randomString(len: Int, upper: Int, lower: Int, num: Int, sym: Int): String {
        var up: String? = null
        if (upper != 0) {
            up = AZ
        }
        var low: String? = null
        if (lower != 0) {
            low = az
        }
        var numerical: String? = null
        if (num != 0) {
            numerical = numeric
        }
        var symbolical: String? = null
        if (sym != 0) {
            symbolical = symbols
        }
        val mapLength = upper + lower + num + sym
        val map = HashMap<Int, String>()

        for (i in 0 until mapLength) {
            if (up != null && !map.containsValue(up)) {
                map[i] = up
            }
            if (low != null && !map.containsValue(low)) {
                map[i] = low
            }
            if (numerical != null && !map.containsValue(numerical)) {
                map[i] = numerical
            }
            if (symbolical != null && !map.containsValue(symbolical)) {
                map[i] = symbolical
            }
        }
        val sb = StringBuilder(len)
        for (i in 0 until len) {
            val type = Random()
            val next = type.nextInt(mapLength)
            val rand = map[next]
            sb.append(rand!![rnd.nextInt(rand.length)])
        }
        return sb.toString()
    }

    override fun onClick(v: View) {
        when (v.id) {
            R.id.genCancel -> finish()
            R.id.genButton -> if (preGenerateCheck()) {
                startGenerating()
            }
            R.id.genInsert -> {
                val intent = Intent()
                val genPass = binding!!.passwordShow.text!!.toString().trim { it <= ' ' }
                intent.putExtra("GENERATED_PASSWORD", genPass)
                setResult(Activity.RESULT_OK, intent)
                finish()
            }
        }
    }

    private fun preGenerateCheck(): Boolean {
        val status: Boolean
        if (binding!!.aZCheck.isChecked || binding!!.azCheck.isChecked || binding!!.numericCheck.isChecked || binding!!.symbolCheck.isChecked) {
            if (binding!!.passwordLength.progress > 0) {
                status = true
            } else {
                status = false
                Toast.makeText(this@GeneratePassword, getString(R.string.pass_length), Toast.LENGTH_SHORT).show()
            }
        } else {
            status = false
            binding!!.aZCheck.isChecked = true
            Toast.makeText(this@GeneratePassword, getString(R.string.pass_checks), Toast.LENGTH_SHORT).show()
        }
        return status
    }

    private fun startGenerating() {
        var aZCh = 0
        if (binding!!.aZCheck.isChecked) {
            aZCh = 1
        }
        var azCh = 0
        if (binding!!.azCheck.isChecked) {
            azCh = 1
        }
        var numCh = 0
        if (binding!!.numericCheck.isChecked) {
            numCh = 1
        }
        var symCh = 0
        if (binding!!.symbolCheck.isChecked) {
            symCh = 1
        }
        val password = randomString(binding!!.passwordLength.progress, aZCh, azCh, numCh, symCh)
        binding!!.passwordShow.setText(password)
    }

    companion object {

        private val AZ = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
        private val az = "abcdefghijklmnopqrstuvwxyz"
        private val numeric = "0123456789"
        private val symbols = "!@#$%()?^*"
    }
}
